docs/static_site/src/pages/api/cpp/docs/tutorials/multi_threaded_inference.md

```diff
@@ -0,0 +1,324 @@
+$ make
+```
+
+If you have built mxnet from source with cmake, please uncomment the specific lines for cmake build or set the following environment variables: `MKLDNN_BUILD_DIR (default is $(MXNET_ROOT)/3rdparty/mkldnn/build)`, `MKLDNN_INCLUDE_DIR (default is $(MXNET_ROOT)/3rdparty/mkldnn/include)`, `MXNET_LIB_DIR (default is $(MXNET_ROOT)/lib)`.
+
+### Download the model and run multi threaded inference example
+To download a model use the `get_model.py` script. This downloads a model to run inference.
+
+```python
+python3 get_model.py --model <model_name>
+```
+e.g.
+```python
+python3 get_model.py --model imagenet1k-inception-bn
+```
+Only the supported models with `get_model.py` work with multi threaded inference.
+
+To run the multi threaded inference example:
+
+First export `LD_LIBRARY_PATH`:
+
+```bash
+$ export LD_LIBRARY_PATH=<MXNET_LIB_DIR>:$LD_LIBRARY_PATH
+```
+
+```bash
+$ ./multi_threaded_inference [model_name] [num_threads] [is_gpu] [file_names]
+```
+e.g.
+
+```bash
+./multi_threaded_inference imagenet1k-inception-bn 2 1 grace_hopper.jpg dog.jpg
+```
+
+The above script spawns 2 threads, shares the same cachedop and params among two threads, and runs inference on GPU. It returns the inference results in the order in which files are provided.
+
+NOTE: This example is to demonstrate the multi-threaded-inference with cached op. The inference results work well only with specific models (e.g. imagenet1k-inception-bn). The results may not necessarily be very accurate because of different preprocessing step required etc.
+
+### Code walkthrough multi-threaded inference with CachedOp
+
+The multi threaded inference example (`multi_threaded_inference.cc`) involves the following steps:
+
+1. Parse arguments and load input image into ndarray
+2. Prepare input data and load parameters, copying data to a specific context
+3. Preparing arguments to pass to the CachedOp and calling C API to **create cached op**
+4. Prepare lambda function which will run in spawned threads. Call C API to **invoke cached op** within the lambda function.
+5. Spawn multiple threads and wait for all threads to complete.
+6. Post process data to obtain inference results and cleanup.
+
+### Step 1: Parse arguments and load input image into ndarray
+
+```c++
+int main(int argc, char *argv[]) {
+  if (argc < 5) {
+    std::cout << "Please provide a model name, num_threads, is_gpu, test_image" << std::endl
+              << "Usage: ./multi_threaded_inference [model_name] [num_threads] [is_gpu] apple.jpg"
+              << std::endl
+              << "Example: ./.multi_threaded_inference imagenet1k-inception-bn 1 0 apple.jpg"
+              << std::endl
+              << "NOTE: Thread number ordering will be based on the ordering of file inputs" << std::endl
+              << "NOTE: Epoch is assumed to be 0" << std::endl;
+    return EXIT_FAILURE;
+  }
+  std::string model_name = std::string(argv[1]);
+  int num_threads = std::atoi(argv[2]);
```

Comment:
let's use std::to_string instead of unsafe atoi in general. This should be a linter rule.

Reply:
If you are hinting at the undefined behavior that will happen only for outside int ranges. As this is for num_threads for an example script this is unlikely. Otherwise, I think atoi seems fine for the use case here.